compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/TailRecursionCallsCollector.kt

```diff
@@ -131,25 +146,25 @@ fun collectTailRecursionCalls(irFunction: IrFunction): Set<IrCall> {
             }
 
             result.add(expression)
-
         }
 
-    }
-
-    val body = irFunction.body
-    if (body !is IrBlockBody) {
-        return emptySet() // TODO: should an assert be here instead?
```

Comment:
@pyos is it possible to add test with expression body?

Reply:
I'm not sure. A single-statement block body is pretty much equivalent to an expression body, so it's really hard to tell in blackbox tests which one is used. I think `fun f() = something` is converted into a block body by psi2ir.

Reply:
Looks like not: https://github.com/JetBrains/kotlin/blob/4752924b1316e65552d42b4ac6af0431cc159085/compiler/testData/ir/irText/declarations/parameters/useNextParamInLambda.txt#L4

NB: I've meant smth like `tailrec fun foo(x: Int): Int = if (x <= 1) 0 else foo(x - 1)`

Reply:
`tailrec fun foo(x: Int): Int = if (x <= 1) 0 else foo(x - 1)` has a block body:

```
    FUN name:foo visibility:public modality:FINAL <> (x:kotlin.Int) returnType:kotlin.Int [tailrec]
      VALUE_PARAMETER name:x index:0 type:kotlin.Int
      BLOCK_BODY
        RETURN type=kotlin.Nothing from='public final fun foo (x: kotlin.Int): kotlin.Int [tailrec] declared in <root>'
          WHEN type=kotlin.Int origin=IF
            ...
```

The line you linked is a default value of a parameter.

Reply:
Also single-line functions like that already appear e.g. in kt36853_fibonacci.